pkg/webhook/resources/setting/validator.go

```diff
@@ -260,8 +260,11 @@ func (v *settingValidator) validateHTTPProxy(setting *v1beta1.Setting) error {
 		}
 	}
 
-	if err := validateHTTPProxyHelper(setting.Value, nodes); err != nil {
-		return werror.NewInvalidError(err.Error(), "value")
+	// Validate the value only if it is not the default value.
```

Comment:
Looks the current code does not handle `nil` well. Both "" and "{}" need to be cared.

The `value` is initialized as "", and `default` as "{}".

The controller side which utilize this setting, should also tolerant  `""`, the `default` can also be an empty string.

```
func validateHTTPProxyHelper(value string, nodes []*corev1.Node) error {
	if value == "" || if value == "{}" {
		return nil
	}

	httpProxyConfig := &util.HTTPProxyConfig{}
	if err := json.Unmarshal([]byte(value), httpProxyConfig); err != nil {
		return err
	}
	// Make sure the node's IP addresses is set in 'noProxy'. These IP
	// addresses can be specified individually or via CIDR address.
	err := validateNoProxy(httpProxyConfig.NoProxy, nodes)
	if err != nil {
		return err
	}
	return nil
}

func (v *settingValidator) validateHTTPProxy(setting *v1beta1.Setting) error {
	nodes, err := v.nodeCache.List(labels.Everything())
	if err != nil {
		return werror.NewInternalError(err.Error())
	}

	if err := validateHTTPProxyHelper(setting.Default, nodes); err != nil {
		return werror.NewInvalidError(err.Error(), "default")
	}

	if err := validateHTTPProxyHelper(setting.Value, nodes); err != nil {
		return werror.NewInvalidError(err.Error(), "value")
	}

	return nil
}
```

Reply:
Your suggestion does not work because it does not abort with an error for a `Value` with `{}`, since in this case `noProxy` does not meet the conditions that the node IPs must be specified.

This happens because in the case of 
```
func validateHTTPProxyHelper(value string, nodes []*corev1.Node) error {
	if value == "" || value == "{}" {
		return nil
	}
	...
```
the function immediately returns with `nil` instead of raising an error at
```
	err := validateNoProxy(httpProxyConfig.NoProxy, nodes)
	if err != nil {
		return err
	}
```

I see another problem, but it already exists before my changes. If `Default` and `Value` are an empty string no error is thrown. This is actually wrong because it is not a valid JSON. And since this setting contains JSON code, an error should be thrown if it is not. Or am I overlooking a general rule here that affects the implementation of settings and their behavior?

Reply:
> When `func (h *Handler) syncHTTPProxy(setting *harvesterv1.Setting) error {` treats `""` as `"{}"`, it will be more robust.
> 
> The setting itself is deined as `string`,`""` is the initial value; user may also delete the validator hook to forcely update value/default as `""`.

I will open a new issue for that.

Reply:
The related enhancement issue is https://github.com/harvester/harvester/issues/6473